### PR TITLE
Repo-policy-compliance checks for OpenStack

### PR DESCRIPTION
**Pull request by cbartz**
### Overview

Add support for an external repo policy compliance web service via the config option. This is currently only used for the Openstack cloud.

Integration tests will be added in a subsequent PR.

### Rationale

Repo policy compliance checks are currently missing for Openstack Cloud.

### Juju Events Changes

n/a

### Module Changes

`charm_state`: Add `RepoPolicyComplianceConfig` and validity checks
`repo_policy_compliance_client`: Move session creation to this module
`runner`: Small adaption to pass url instead of host-ip to pre-job content
`runner_manager`: Remove session creation
`pre-job.j2`: Use url instead of host-ip parameter

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [ ] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)
Documentation on charmhub will be added once openstack feature is ready.

### CODE REVIEW

File: src/charm_state.py [hunk 1]
```diff
@@ -337,8 +381,9 @@ def _parse_dockerhub_mirror(cls, charm: CharmBase) -> str | None:
         if not dockerhub_mirror:
             return None
 
+        dockerhub_mirror = cast(str, dockerhub_mirror)
         dockerhub_mirror_url = urlsplit(dockerhub_mirror)
-        if dockerhub_mirror is not None and dockerhub_mirror_url.scheme != "https":
+        if dockerhub_mirror_url.scheme != "https":
```
Comment (yanksyoon) [inline src/charm_state.py:386]: Out of curiosity, would we be able to leverage HttpUrl with Pydantic and `allowed_schemes` set to `https` for this check?
https://docs.pydantic.dev/latest/api/networks/#pydantic.networks.HttpUrl

(Out of scope for this PR so just leaving a comment, feel free to resolve!)

**Review by yanksyoon — COMMENTED** [review]
Reviewed up to pre-job.j2, LGTM so far!
Thanks for fixing minor linting issues as well!

File: config.yaml [hunk 2]
```diff
@@ -86,6 +86,18 @@ options:
...
+    description: >-
+      The token to authenticate with the repository-policy-compliance service in order to 
+      generate one-time-tokens. This option requires the repo-policy-compliance-url to be set.
+      If not set, the repository-policy-compliance service will not be used.
+  repo-policy-compliance-url:
+    type: string
+    description: >-
+      The URL to the repository-policy-compliance service. This option requires the 
+      repo-policy-compliance-token to be set. If not set, the repository-policy-compliance service
+      will not be used. This option is currently only effective when using OpenStack Cloud.
```
Comment (jdkandersson) [inline config.yaml:100]: Should these put the charm in blocked state if, for example, only 1 is set or they are set when not using OpenStack?

Comment (cbartz) [reply config.yaml]: Have updated to go into blocked state when this option is used without OpenStack (code is in `charm_state`)

File: src/charm_state.py (hunk 1 above)
Comment (cbartz) [reply src/charm_state.py]: Updated

**Review by yanksyoon — APPROVED** [review]
LGTM!

**Pull request merged**